system/include/emscripten/dom_pk_codes.h

```diff
@@ -0,0 +1,506 @@
+#define DOM_PK_EJECT                0xE02C /* "Eject"              */
+#define DOM_PK_AUDIO_VOLUME_DOWN    0xE02E /* "AudioVolumeDown"    */
+#define DOM_PK_AUDIO_VOLUME_DOWN    0xE02E /* "VolumeDown"         */
+#define DOM_PK_AUDIO_VOLUME_UP      0xE030 /* "AudioVolumeUp"      */
+#define DOM_PK_AUDIO_VOLUME_UP      0xE030 /* "VolumeUp"           */
+#define DOM_PK_BROWSER_HOME         0xE032 /* "BrowserHome"        */
+#define DOM_PK_NUMPAD_DIVIDE        0xE035 /* "NumpadDivide"       */
+#define DOM_PK_ALT_RIGHT            0xE038 /* "AltRight"           */
+#define DOM_PK_HELP                 0xE03B /* "Help"               */
+#define DOM_PK_NUM_LOCK             0xE045 /* "NumLock"            */
+#define DOM_PK_HOME                 0xE047 /* "Home"               */
+#define DOM_PK_ARROW_UP             0xE048 /* "ArrowUp"            */
+#define DOM_PK_PAGE_UP              0xE049 /* "PageUp"             */
+#define DOM_PK_ARROW_LEFT           0xE04B /* "ArrowLeft"          */
+#define DOM_PK_ARROW_RIGHT          0xE04D /* "ArrowRight"         */
+#define DOM_PK_END                  0xE04F /* "End"                */
+#define DOM_PK_ARROW_DOWN           0xE050 /* "ArrowDown"          */
+#define DOM_PK_PAGE_DOWN            0xE051 /* "PageDown"           */
+#define DOM_PK_INSERT               0xE052 /* "Insert"             */
+#define DOM_PK_DELETE               0xE053 /* "Delete"             */
+#define DOM_PK_META_LEFT            0xE05B /* "MetaLeft"           */
+#define DOM_PK_OS_LEFT              0xE05B /* "OSLeft"             */
+#define DOM_PK_META_RIGHT           0xE05C /* "MetaRight"          */
+#define DOM_PK_OS_RIGHT             0xE05C /* "OSRight"            */
+#define DOM_PK_CONTEXT_MENU         0xE05D /* "ContextMenu"        */
+#define DOM_PK_POWER                0xE05E /* "Power"              */
+#define DOM_PK_BROWSER_SEARCH       0xE065 /* "BrowserSearch"      */
+#define DOM_PK_BROWSER_FAVORITES    0xE066 /* "BrowserFavorites"   */
+#define DOM_PK_BROWSER_REFRESH      0xE067 /* "BrowserRefresh"     */
+#define DOM_PK_BROWSER_STOP         0xE068 /* "BrowserStop"        */
+#define DOM_PK_BROWSER_FORWARD      0xE069 /* "BrowserForward"     */
+#define DOM_PK_BROWSER_BACK         0xE06A /* "BrowserBack"        */
+#define DOM_PK_LAUNCH_APP_1         0xE06B /* "LaunchApp1"         */
+#define DOM_PK_LAUNCH_MAIL          0xE06C /* "LaunchMail"         */
+#define DOM_PK_LAUNCH_MEDIA_PLAYER  0xE06D /* "LaunchMediaPlayer"  */
+#define DOM_PK_MEDIA_SELECT         0xE06D /* "MediaSelect"        */
+
+/* Maps the EmscriptenKeyboardEvent::code field from emscripten/html5.h to one of the DOM_PK codes above. */
+static DOM_PK_CODE_TYPE emscripten_compute_dom_pk_code(const char *keyCodeString)
+{
+  if (!keyCodeString) return 0;
+
+  /* Compute the collision free hash. */
+  unsigned int hash = 0;
+  while(*keyCodeString) hash = ((hash ^ 0x5EE148E8U) << 3) ^ (unsigned int)*keyCodeString++;
```

Comment:
> maybe i am missing the point here. but the hashing seems like overkill, we could just do strcmp? (switching on first character etc. for speed)?

I thought about `switch-case`ing on strings per character, but that implementation would look quite bad, since it would be lots of nested switching, and implementing such a `switch-case` network for given list of fixed input strings is a task for a script generator as well. One could switch-case on first few characters and then fall back to linear `strcmp` for the remaining ones, but that seems much more complex to read and debug than this.

Another benefit of hashing over a switch-case network is smaller compiled code size. Hashing needs `sizeof(unsigned int)*num_pk_codes` amount of data in the compiled code, whereas a string based solution would need to store all the actual strings of all the different possible physical `code` identifiers in the compiled output. (minus some possible overlap)

Hashing here is not complex, the function is implemented as a one liner, the only complexity is the script that searches for the collision free function, though that is not too long anyways.

> also, does the perfect hash here prevent a string not from this set (i.e. an invalid input) from hitting a valid hash and being incorrectly accepted?

Yes, though I think that is perfectly acceptable, since the set of strings is predetermined by the documented standard for the `code` field, and if there are any new strings that are added by browsers, those can be added and the script rerun to rebuild the header. The PK enums cannot support unidentified keys no matter what, and having a collision on unidentified key is not particularly critical, that will just mean that this new key that browsers added support for would produce the same input in the application as an existing key. Most likely the new key is not identified and a `DOM_PK_UNKNOWN` is returned, prompting us to add support.